File: frameworks/python-playwright/services/order_service.py
"""Builds the controlled order several extended scenarios need as a precondition.

Mirrors `services/order_service.ts`: a private product, one cart unit and one
order whose financial values are reconciled with the platform before the
assertions of the test start.
"""

from __future__ import annotations

import time
from typing import Any

from data.test_data import PRODUCT
from helpers.api_response import expect_api_json
from helpers.unique_name import unique_product_name
from services.api_client import ApiClient
from services.product_service import ProductService


class ControlledOrder:
    def __init__(self, order: dict[str, Any], product: dict[str, Any]) -> None:
        self.order = order
        self.product = product


class OrderService:
    def __init__(self, api: ApiClient, products: ProductService) -> None:
        self.api = api
        self.products = products

    def create_controlled_order(self, payment_status: str = "paid") -> ControlledOrder:
        self.products.clear_cart()
        product = self.products.create_product({"name": unique_product_name(), **PRODUCT})

        added = self.api.post(
            "/api/cart", json={"product_id": product["id"], "quantity": 1}
        )
        assert added.status_code == 201, added.text

        shipping = {
            "full_name": "Framework Test User",
            "email": f"framework.{time.time_ns()}@example.test",
            "address_line": "123 Main Street",
            "city": "New York",
            "state": "NY",
            "zip_code": "10001",
            "country": "USA",
            "is_fake": True,
        }
        payment = {
            "card_number": "4242424242424242",
            "expiry_date": "12/25",
            "cvv": "123",
            "brand": "Visa",
            "payment_status": payment_status,
        }
        financial = {"tax": 0, "shipping": 0, "state": "NY", "country": "USA"}

        response = self.api.post(
            "/api/orders",
            json={"shipping": shipping, "payment": payment, "financial": financial},
        )
        if response.status_code == 400:
            # The platform recalculates tax and shipping and answers SYNC_ERROR
            # with the values it expects, so the order is retried once with them.
            mismatch = expect_api_json(response)
            assert mismatch.get("error") == "SYNC_ERROR", response.text
            financial = {
                "tax": mismatch["correct_values"]["tax"],
                "shipping": mismatch["correct_values"]["shipping_cost"],
                "state": "NY",
                "country": "USA",
            }
            response = self.api.post(
                "/api/orders",
                json={"shipping": shipping, "payment": payment, "financial": financial},
            )

        assert response.status_code == 201, response.text
        created = expect_api_json(response)["data"]
        order = {**created, "id": created["order_id"]}
        return ControlledOrder(order, product)

    def get(self, order_id: int) -> dict[str, Any]:
        response = self.api.get(f"/api/orders/{order_id}")
        assert response.status_code == 200, response.text
        return expect_api_json(response)["data"]

    def cleanup(
        self, controlled: ControlledOrder | None, order_already_deleted: bool = False
    ) -> None:
        if controlled is None:
            return
        if not order_already_deleted:
            self.products.delete_order(controlled.order["id"])
        self.products.clear_cart()
        self.products.delete_product(controlled.product["id"])

File: frameworks/typescript-cypress/services/order_service.ts
import { teachingData } from '../data/test_data';
import { guardedApi } from '../helpers/api_response';
import { uniqueProductName } from '../helpers/unique_name';
import { ProductService, authHeaders, type Product } from './product_service';

export type Order = {
  id: number;
  subtotal: number;
  tax: number;
  shipping: number;
  total: number;
  status: string;
  payment_status: string;
  order_items?: Array<Record<string, unknown>>;
  order_shipping?: Record<string, unknown> | Array<Record<string, unknown>> | null;
};

export type ControlledOrder = { order: Order; product: Product };

type Financial = { tax: number; shipping: number; state: string; country: string };

/**
 * Creates and removes the paid orders the extended catalog needs. Orders are the
 * one resource the platform will not let a test invent, so every scenario that
 * reads an order builds its own and deletes it again.
 */
export class OrderService {
  constructor(private readonly products = new ProductService()) {}

  createControlledOrder(tag: string, paymentStatus = 'paid'): Cypress.Chainable<ControlledOrder> {
    this.products.clearCart();
    return this.products
      .createProduct({ name: uniqueProductName(tag), ...teachingData.product })
      .then((product) =>
        guardedApi({
          method: 'POST',
          url: '/api/cart',
          headers: authHeaders(),
          body: { product_id: product.id, quantity: 1 },
        })
          .then((added) => {
            expect(added.status, 'add to cart status').to.eq(201);
            return this.placeOrder(paymentStatus);
          })
          .then((order) => ({ order, product })),
      );
  }

  get(id: number): Cypress.Chainable<Order> {
    return guardedApi<{ data: Order }>({
      url: `/api/orders/${id}`,
      headers: authHeaders(),
    }).then((response) => {
      expect(response.status, 'read order status').to.eq(200);
      return response.body.data;
    });
  }

  cleanup(controlled?: ControlledOrder, orderAlreadyDeleted = false): void {
    if (!controlled) return;
    if (!orderAlreadyDeleted) this.products.deleteOrder(controlled.order.id);
    this.products.clearCart();
    this.products.deleteProduct(controlled.product.id);
  }

  /**
   * The platform recomputes tax and shipping server side and rejects a mismatch
   * with SYNC_ERROR plus the values it expects, so the first rejection is used
   * to resubmit rather than to fail.
   */
  private placeOrder(paymentStatus: string): Cypress.Chainable<Order> {
    const shipping = {
      full_name: 'Framework Test User',
      email: `framework.${Date.now()}@example.test`,
      address_line: '123 Main Street',
      city: 'New York',
      state: 'NY',
      zip_code: '10001',
      country: 'USA',
      is_fake: true,
    };
    const payment = {
      card_number: '4242424242424242',
      expiry_date: '12/25',
      cvv: '123',
      brand: 'Visa',
      payment_status: paymentStatus,
    };
    const submit = (financial: Financial) =>
      guardedApi<any>({
        method: 'POST',
        url: '/api/orders',
        headers: authHeaders(),
        body: { shipping, payment, financial },
      });

    return submit({ tax: 0, shipping: 0, state: 'NY', country: 'USA' })
      .then((response) => {
        if (response.status !== 400) return cy.wrap(response, { log: false });
        expect(response.body.error, 'financial mismatch error').to.eq('SYNC_ERROR');
        return submit({
          tax: response.body.correct_values.tax,
          shipping: response.body.correct_values.shipping_cost,
          state: 'NY',
          country: 'USA',
        });
      })
      .then((response: any) => {
        expect(response.status, JSON.stringify(response.body)).to.eq(201);
        const created = response.body.data as Order & { order_id: number };
        // The creation payload names the identifier `order_id`; every later read
        // uses `id`, so the service normalizes it once here.
        const order: Order = { ...created, id: created.order_id };
        return order;
      });
  }
}
